File: heap/Solution313.java
import java.util.Set;
import java.util.HashSet;
import java.util.Queue;
import java.util.PriorityQueue;

class Solution313 {
    public int nthSuperUglyNumber(int n, int[] primes) {
    	Set<Long> set = new HashSet<>();
    	Queue<Long> pq = new PriorityQueue<>();
    	int index = 0;
    	pq.add((long) 1);
    	set.add((long) 1);

    	while(index < n){
    		long cur = pq.poll();
    		index += 1;
    		if(index == n){
    			return (int) cur;
			}
    		for(int j = 0; j < primes.length; j++) {
				if (!set.contains(cur * primes[j])) {
					set.add(cur * primes[j]);
					pq.add(cur * primes[j]);
				}
			}
    	}
    	return -1;
    }

    public static void main(String[] args){
    	Solution313 s = new Solution313();
    	System.out.println(s.nthSuperUglyNumber(12, new int[]{2, 7, 13, 19}));
	}
}

File: C++/Solution313.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<queue>
using namespace std;

typedef long long ll;

class Solution313 {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {
        priority_queue<ll, vector<ll>, greater<ll>> pq;
        unordered_set<ll> st;
        pq.push(1);
        st.insert(1);
        ll cur = 0;
        while(n > 0){
            cur = pq.top();
            pq.pop();
            for(int prime: primes){
                ll x = cur * prime;
                if(!st.count(x)){
                    pq.push(x);
                    st.insert(x);
                }
            }
            n --;
        }
        return cur;
    }
};